chain/network/src/test_loop.rs

```diff
@@ -1,4 +1,301 @@
+pub struct ClientSenderForTestLoopNetwork {
+    pub block: AsyncSender<BlockResponse, ()>,
+    pub block_approval: AsyncSender<BlockApproval, ()>,
+    pub transaction: AsyncSender<ProcessTxRequest, ProcessTxResponse>,
+    pub chunk_endorsement: AsyncSender<ChunkEndorsementMessage, ()>,
+}
+
+type NetworkRequestHandler = Arc<dyn Fn(NetworkRequests) -> Option<NetworkRequests>>;
+
+/// A custom actor for the TestLoop framework that can be used to send network messages across clients
+/// in a multi-node test.
+///
+/// This actor has a set of handlers to handle PeerManagerMessageRequest messages. We have a set of
+/// default handlers that handle messages sent to client, partial_witness actor, and shards_manager.
+/// It is possible to override these handlers by registering a new handler using the
+/// `register_override_handler()` method.
+///
+/// The signature of the handler is `dyn Fn(NetworkRequests) -> Option<NetworkRequests>`.
+/// If the handler returns None, it means that the message was handled and no further processing is
+/// required. If the handler returns Some(request), it means that the message was not handled and
+/// the request should be passed to the next handler in the chain.
+///
+/// It's possible for a handler to modify the data in request and return it. This can be useful for
+/// simulating things like malicious actors where we can modify the data in the request.
+///
+/// In case no handler is able to handle the request, the actor will panic.
+///
+/// NOTE: To make the override functionality work with the default handlers, the handlers are tried in
+/// reverse order.
+///
+/// Examples of custom handlers
+/// - Override handler to skip sending messages to or from a specific client.
+/// - Override handler to simulate more network delays.
+/// - Override handler to modify data and simulate malicious behavior.
+#[derive(Default)]
+pub struct TestLoopPeerManagerActor {
+    handlers: Vec<NetworkRequestHandler>,
+}
+
+impl Actor for TestLoopPeerManagerActor {}
+
+impl TestLoopPeerManagerActor {
+    /// Create a new TestLoopPeerManagerActor with default handlers for client, partial_witness, and shards_manager.
+    /// Note that we should be able to access the senders for these actors from the data type.
+    pub fn new<'a, T>(clock: Clock, account_id: &AccountId, datas: &'a Vec<T>) -> Self
+    where
+        AccountId: From<&'a T>,
+        ClientSenderForTestLoopNetwork: From<&'a T>,
+        PartialWitnessSenderForNetwork: From<&'a T>,
+        Sender<ShardsManagerRequestFromNetwork>: From<&'a T>,
+    {
+        let mut actor = Self { handlers: Vec::new() };
+        vec![
+            network_message_to_client_handler(&account_id, make_sender_map(datas)),
+            network_message_to_partial_witness_handler(&account_id, make_sender_map(datas)),
+            network_message_to_shards_manager_handler(clock, &account_id, make_sender_map(datas)),
+            network_message_to_state_snapshot_handler(),
```

Comment:
That's out of scope of the PR, but if current NetworkRequest handling implementation is equivalent to calling all handlers one-by-one, we should make it an enum:
```
NetworkRequest::ClientRequest
NetworkRequest::PartialWitnessActorRequest
NetworkRequest::ShardsManagerActorRequest
...
```